http/events.go

```diff
@@ -69,14 +87,80 @@ func handleEventsSubscribeWebsocket(args eventSubscribeArgs) (websocket.StatusCo
 				logger.Warn("Could not serialize websocket event", "error", err)
 				return 0, "", err
 			}
-			err = args.conn.Write(ctx, messageType, messageBytes)
+			err = sub.conn.Write(ctx, messageType, messageBytes)
 			if err != nil {
 				return 0, "", err
 			}
 		}
 	}
 }
 
+// allowMessageCached checks that the message is allowed to received by the websocket.
+// It caches results for specific namespaces, data paths, and event types.
+func (sub *eventSubscriber) allowMessageCached(message *logical.EventReceived) bool {
+	// default to the subscribe endpoint if there is no specific data path
+	messageNs := strings.Trim(message.Namespace, "/")
+	dataPath := "sys/events/subscribe"
+	if message.Event.Metadata != nil {
+		dataPathField := message.Event.Metadata.GetFields()[logical.EventMetadataDataPath]
+		if dataPathField != nil {
+			dataPath = dataPathField.GetStringValue()
+		}
+	}
+	cacheKey := fmt.Sprintf("%v!%v!%v", messageNs, dataPath, message.EventType)
+	_, ok := sub.checkCache.Get(cacheKey)
+	if ok {
+		return true
+	}
+
+	// perform the actual check and cache it if true
+	ok = sub.allowMessage(messageNs, dataPath, message.EventType)
+	if ok {
+		err := sub.checkCache.Add(cacheKey, ok, webSocketRevalidationTime)
```

Comment:
nit: There is a race here between checking the cache and adding a result to it. Ideally, we could have some sort of atomic `GetOrAdd` function that gets passed a function for calculating the result on a cache miss.

Reply:
We only add in the case if the result is true. So, there's a race, but there will never be conflicting values added.

I don't know if it is worth adding the extra time to eliminate the race condition?

Reply:
Yeah agreed - I think ignore this one. Given the cache is per-token and the reevaluation period is 5 minutes, the chances of an actual race occurring are tiny right now anyway.